[Term Entry] Java Strings: .substring() (#1565)
Codecademy/docs: content/java/concepts/strings/terms/substring/substring.md

---
Title: '.substring()'
Description: 'Returns a part of the string specified through a starting index and an optional ending index.'
Subjects:
  - 'Computer Science'
Tags:
  - 'Characters'
  - 'Strings'
  - 'Methods'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
---

> The **`.substring()`** method returns a part of the string specified through a starting index and an optional ending index. If the ending index is not provided, then the substring will run from the starting index to the end of the original string.

Reviewer: ```suggestion
The **`.substring()`** method returns a part of the string from some starting index to an (optional) ending index. If the ending index is not provided, the substring will start from the starting index to the end of the original string.
```

## Syntax

```pseudo
string.substring(startIndex);
string.substring(startIndex, endIndex);
```

> - `.substring()` returns characters from start index up to, but not including, the character at the end index.

Reviewer: ```suggestion
`.substring()` returns characters from `startIndex` up to, but not including, the character at `endIndex`.
```

> - If the end index is omitted `.substring()` returns characters from the start index to the end of the string.

Reviewer: ```suggestion
If `endIndex` is omitted, `.substring()` returns characters from `startIndex` to the end of the string.
```

> - If the start and end indexes are equal, `.substring()` returns an empty string.

Reviewer: ```suggestion
If `startIndex` and `endIndex` are equal, `.substring()` returns an empty string.
```

> - If `startIndex` is greater than the `endIndex`, greater than the length of the sequence, or if either index is negative then a [`StringIndexOutOfBoundsException`](https://www.codecademy.com/resources/docs/java/errors/stringindexoutofboundsexception) is thrown.

Reviewer: ```suggestion
A [`StringIndexOutOfBoundsException`](https://www.codecademy.com/resources/docs/java/errors/stringindexoutofboundsexception) is thrown if any of the following are true:
- `startIndex` is greater than  `endIndex`
- `startIndex` is greater than the length of the original string
- `startIndex` or `endIndex` is negative
```

## Example

> The following example uses `.substring()` to display a subset of characters from a given string.

Reviewer: ```suggestion
The following example uses `.substring()` to display a substring of a given string.
```

---
Title: '.substring()'
Description: 'Returns a part of the string specified through starting and an optional ending index. If ending index is not specified, then the substring expands to the end of original string.'
Subjects:
  - 'Computer Science'
Tags:
  - 'Characters'
  - 'Strings'
  - 'Methods'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
---

The **`.substring()`** method returns a part of the string specified through starting and an optional ending index. If the ending index is not provided, then the substring expands to the end of of original string.

## Syntax

```pseudo
// Returns substring from startIndex to the end of string

> string.substring(int startIndex);

Author: This line should not be removed, because it gives an example of syntax with only starting index. This is also done for current `.substring()` term for JavaScript.

// Returns substring from startIndex to (endIndex - 1)
string.substring(int startIndex, int endIndex);
```

## Details

- `.substring()` returns characters from start index up to, but not including, the character at the end index.

> - If the end index is omitted `.substring()` returns characters from the start index up through the end of the string.

Reviewer 2: ```suggestion
- If the end index is omitted `.substring()` returns characters from the start index to the end of the string.
```

- If the start and end indexes are equal, `.substring()` returns an empty string.

> - If `startIndex` is negative, greater than the `end`, or greater than the length of the sequence then a [`StringIndexOutOfBoundsException`](https://www.codecademy.com/resources/docs/java/errors/stringindexoutofboundsexception) is thrown.

Reviewer 2: ```suggestion
- If `startIndex` is greater than the `endIndex`,  greater than the length of the sequence, or if either index is negative then a [`StringIndexOutOfBoundsException`](https://www.codecademy.com/resources/docs/java/errors/stringindexoutofboundsexception) is thrown.
```

## Example

> Using `.substring()` to display characters from a given string.

Reviewer 2: ```suggestion
The following example uses `.substring()` to display a subset of characters from a given string.
```

> ```java

Reviewer 2: ```suggestion
```java
// Example.java
```

public class Example {
  public static void main(String[] args) {

>     String str = "Java is cool!";

Reviewer 2: ```suggestion
    String blurb = "Java is cool!";
```

>     // printing last 5 characters of str

Reviewer 2: ```suggestion
    // Printing last 5 characters of blurb
```

>     System.out.println(str.substring(str.length() - 5));

Reviewer 2: ```suggestion
    System.out.println(blurb.substring(blurb.length() - 5));
```

>     // printing first 4 characters of str

Reviewer 2: ```suggestion
    // Printing first 4 characters of blurb
```

>     System.out.println(str.substring(0, 5));

Reviewer 2: ```suggestion
    System.out.println(blurb.substring(0, 5));
```
